File: src/main/java/io/scriptor/machine/Semihosting.java
package io.scriptor.machine;

import org.jetbrains.annotations.NotNull;

public interface Semihosting {

    int SEMIHOSTING_SYSOPEN = 0x01;
    int SEMIHOSTING_SYSWRITEC = 0x03;
    int SEMIHOSTING_SYSWRITE = 0x05;
    int SEMIHOSTING_SYSREAD = 0x06;
    int SEMIHOSTING_SYSREADC = 0x07;
    int SEMIHOSTING_SYSERRNO = 0x13;

    int SEMIHOSTING_SYSOPEN_MODE_READ = 0x0;
    int SEMIHOSTING_SYSOPEN_MODE_BINARY = 0x1;
    int SEMIHOSTING_SYSOPEN_MODE_PLUS = 0x2;
    int SEMIHOSTING_SYSOPEN_MODE_WRITE = 0x4;
    int SEMIHOSTING_SYSOPEN_MODE_APPEND = 0x8;

    static long fopen(final @NotNull Machine machine, final @NotNull String fname, final long mode, final long len) {
        return -1;
    }

    static long fread(final @NotNull Machine machine, final long fd, final long memp, final long len) {
        return 0L;
    }

    static long fwrite(final @NotNull Machine machine, final long fd, final long memp, final long len) {
        return 0L;
    }
}

File: src/main/java/io/scriptor/machine/Semihosting.kt
package io.scriptor.machine

interface Semihosting {

    companion object {
        fun fopen(machine: Machine, fname: String, mode: ULong, len: ULong): ULong {
            // TODO
            return 0UL.inv()
        }

        fun fread(machine: Machine, fd: ULong, memp: ULong, len: ULong): ULong {
            // TODO
            return 0UL
        }

        fun fwrite(machine: Machine, fd: ULong, memp: ULong, len: ULong): ULong {
            // TODO
            return 0UL
        }

        const val SEMIHOSTING_SYSOPEN = 0x01UL
        const val SEMIHOSTING_SYSWRITEC = 0x03UL
        const val SEMIHOSTING_SYSWRITE = 0x05UL
        const val SEMIHOSTING_SYSREAD = 0x06UL
        const val SEMIHOSTING_SYSREADC = 0x07UL
        const val SEMIHOSTING_SYSERRNO = 0x13UL

        const val SEMIHOSTING_SYSOPEN_MODE_READ = 0x0
        const val SEMIHOSTING_SYSOPEN_MODE_BINARY = 0x1
        const val SEMIHOSTING_SYSOPEN_MODE_PLUS = 0x2
        const val SEMIHOSTING_SYSOPEN_MODE_WRITE = 0x4
        const val SEMIHOSTING_SYSOPEN_MODE_APPEND = 0x8
    }
}
